### ISSUE TITLE
Appendix.C. Ablation of Value Injection

### ISSUE THREAD

**Issue by ghosard** [Shilin-LU/TF-ICON#23]
Thanks for your great work!

I agree to the statement "Injecting all the information might result in a more rigid generation"; and when I try to also inject the values into the  generation steps, the results seem to have a significant gap compared to the figures in the paper.

So I would like to know the detailed settings in your code to inject the values in the generation process, thanks a lot :)

**Comment by Shilin-LU** [comment]
Thank you for your interest!

Injecting both the attention map and value is equivalent to injecting the output feature. Thus, we implement it by directly composing the feature of self-attention layers.

**Comment by ghosard** [comment]
Thanks for your kind reply!
following your comment, I get better results (left), but it's still somehow different from the result shown in the papaer(right).

<img width="200" height="200" alt="image" src="https://github.com/Shilin-LU/TF-ICON/assets/85659543/ed923fb5-20a0-4958-825d-265c31ec8c2b">
<img width="200" height="200" alt="image" src="https://github.com/Shilin-LU/TF-ICON/assets/85659543/29f6a2b6-c9df-4ea9-9e34-3609f3f05e58">

For detail, I store the `out` in ref image's self attention layers; and inject it 
```python
... 
out = torch.cat([out]*2, dim=0)
if encode == False:
    sim = controller(sim, is_cross, place_in_unet, out) # extre store out in v_store
...                
```
then, I inject it in the generation step, using the index calculated above:
```python
...
if place_in_unet == 'down':
    if inject and len(ref_loc_masked)!= 0:
        out[b//2:, orig_loc_masked,:] = controller_for_inject[1].v_store['down_self'][layernum][b//2:, ref_loc_masked, :] # row injection
elif place_in_unet == 'up':
    if inject and len(ref_loc_masked)!= 0:
        out[b//2:, orig_loc_masked,:] = controller_for_inject[1].v_store['up_self'][layernum][b//2:, ref_loc_masked, :] # row injection
elif place_in_unet == 'mid':
    if inject and len(ref_loc_masked)!= 0:
        out[b//2:, orig_loc_masked,:] = controller_for_inject[1].v_store['mid_self'][layernum][b//2:, ref_loc_masked, :] # row injection
...
```
and injecting pesudo cross-attention, ref's attention map,  and background injecting are all **commented out**, for just self attention layer injecting.

Is there still something different from your settings? The hyperparameters are the same from the origin code settings, does it still need cross-attention injecting or something else? Or, are the indices for layer injecting correct?

**Comment by Shilin-LU** [comment]
Oh, for this ablation study, there might be errors in my implementation. Sorry about that. Since I didn't retain that part of the code, I guess it's possible that I mistakenly combined x_t^fg and x_t^bg at certain timesteps according to the mask. You might want to try and see if you can get similar results this way.

**Comment by ghosard** [comment]
Ok, I understand:) 
So the results of combining x_t^fg and x_t^bg correspond to my experiment results.
Exactly, now I'm also working on training-free composition; and the abla results seem very similar to my results.
I'm trying doing something in the latent space; and your abla results, according to the paper, is working on transformer layers rather than directly changing the latent's vectors. That's why I want to reproduce your abla results: different methods (transfomer layers editing and latents editing ) result in almost identical results.
I don't have any more questions now, thank you again for your kind responses:)

**Comment by Shilin-LU** [comment]
> Ok, I understand:) So the results of combining x_t^fg and x_t^bg correspond to my experiment results. Exactly, now I'm also working on training-free composition; and the abla results seem very similar to my results. I'm trying doing something in the latent space; and your abla results, according to the paper, is working on transformer layers rather than directly changing the latent's vectors. That's why I want to reproduce your abla results: different methods (transfomer layers editing and latents editing ) result in almost identical results. I don't have any more questions now, thank you again for your kind responses:)

Awesome, it's exciting to know you're also exploring training-free composition. Wish you the best in your experiments!
